lib/src/rules/close_sinks.dart

```diff
@@ -0,0 +1,137 @@
+import 'package:linter/src/utilities/dart_type_utilities.dart';
+
+typedef void _VisitVariableDeclaration(VariableDeclaration node);
+
+const _desc = r'Close instances of `dart.core.Sink`.';
+
+const _details = r'''
+
+**DO** Invoke `close` on instances of `dart.core.Sink` to avoid memory leaks and
+unexpected behaviors.
+
+**BAD:**
+```
+class A {
+  IOSink _sinkA;
+  void init(filename) {
+    _sinkA = new File(filename).openWrite(); // LINT
+  }
+}
+```
+
+**BAD:**
+```
+void someFunction() {
+  IOSink _sinkF; // LINT
+}
+```
+
+**GOOD:**
+```
+class B {
+  IOSink _sinkB;
+  void init(filename) {
+    _sinkB = new File(filename).openWrite(); // OK
+  }
+
+  void dispose(filename) {
+    _sinkB.close();
+  }
+}
+```
+
+**GOOD:**
+```
+void someFunctionOK() {
+  IOSink _sinkFOK; // OK
+  _sinkFOK.close();
+}
+```
+''';
+
+class CloseSinks extends LintRule {
+  CloseSinks() : super(
+      name: 'close_sinks',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  Visitor(this.rule);
+
+  @override
+  void visitVariableDeclarationStatement(VariableDeclarationStatement node) {
+    FunctionDeclaration function =
+        node.getAncestor((a) => a is FunctionDeclaration);
+    node.variables.variables.forEach(_buildVisitVariableCallback(function));
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    ClassDeclaration classDecl =
+    node.getAncestor((a) => a is ClassDeclaration);
+    node.fields.variables.forEach(_buildVisitVariableCallback(classDecl));
+  }
+
+  _VisitVariableDeclaration _buildVisitVariableCallback(AstNode container) =>
+          (VariableDeclaration sink) {
+        if (!_implementsDartCoreSink(sink.element.type)) {
+          return;
+        }
+
+        List<AstNode> containerNodes = _traverseNodesInDFS(container);
+        Iterable<AstNode> closeSinkInvocations = _findNodesClosingSink(
+            containerNodes, sink);
+
+        if (closeSinkInvocations.isEmpty) {
+          rule.reportLint(sink);
+        }
+      };
+}
+
+Iterable<AstNode> _findNodesClosingSink(List<AstNode> classNodes,
+    VariableDeclaration variable) => classNodes.where(
+    (n) =>
+        n is MethodInvocation &&
+            n.methodName.name == 'close' &&
+            n.target is SimpleIdentifier &&
+            (n.target as SimpleIdentifier).name == variable.name.name);
+
+bool _implementsDartCoreSink(DartType type) =>
+    DartTypeUtilities.findImplementedInterfaces(type).any(_isDartCoreSink);
+
+bool _isDartCoreSink(InterfaceType interface) =>
+    interface.name == 'Sink' &&
+        interface.element.library.name == 'dart.core';
+
+/// Builds the list resulting from traversing the node in DFS and does not
+/// include the node itself.
+List<AstNode> _traverseNodesInDFS(AstNode node) {
```

Comment:
This seems like it would be really inefficient because it's creating and discarding (potentially) large numbers of lists. We took a different approach with NodeLocator that might be worth considering. It would certainly be interesting to do a little performance comparison between the two approaches.

But even if you stick with creating a list, you could create a single list and pass it in to a helper method that does the actual traversal, which would eliminate most of the extra objects. (I still have concerns about building a huge list of AST nodes before starting to look at any of them.)